### PR TITLE
STORM-2208: HDFS State Throws FileNotFoundException in Azure Data Lake Store file system (adl://)

### PR DESCRIPTION
**Pull request by hmcl**
- Close OutputStream before file gets deleted

### CODE REVIEW

**Comment by hmcl** [comment]
Please backport to 1.x and 1.0.x branch. It should be a trivial merge.

**Comment by harshach** [comment]
+1

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/trident/HdfsState.java [hunk 1]
```diff
@@ -500,6 +500,7 @@ private void updateIndex(long txId) {
             bw.write(txnRecord.toString());
             bw.newLine();
             bw.flush();
+            out.close();       // Closing the output stream in finally block causes FileNotFoundException in Azure Data Lake Store file system (adl://)
```
Comment (revans2) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/trident/HdfsState.java]: Why not just wrap the output stream in a try block instead?

```
Path tmpPath = tmpFilePath(indexFilePath.toString());
TxnRecord txnRecord = new TxnRecord(txId, options.currentFile.toString(), this.options.getCurrentOffset());
try (FSDataOutputStream out = this.options.fs.create(tmpPath, true);
       BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(out));) {
    bw.write(txnRecord.toString());
    bw.newLine();
    bw.flush();
}
```

Then you don't need to worry about the close happening before the delete, even in error cases.

**Comment by hmcl** [comment]
@revans2 agree && Done! 

I followed your suggestion but most calls to fs will throw IOException. To keep only one try catch block the cleanest way to do it is to close the OutputStream inside the try block. Otherwise, we need 2 try / catch blocks.

**Comment by revans2** [comment]
+1 looks god to me, although I am not an expert on azure, so I assume you have tested the new code and it works.

**Comment by arunmahadevan** [comment]
+1, the changes look good. Not being able to close the stream after renaming / deleting the file is deviation from the hdfs/java fs standard which should also be supported by the azure implementation.

**Comment by hmcl** [comment]
@arunmahadevan I agree! I have already filed a but to the Azure team to fix that case.

However, it does no harm that we account for this scenario.

**Comment by hmcl** [comment]
@revans2 I just tested in the Azure cluster running AFS and the fix works. Please see attached screenshots.

broken_2016-11-22_6 43 08pm
fixed_2016-11-22_6 34 29pm

**Comment by HeartSaVioR** [comment]
+1

**Pull request merged**